### PR TITLE
9326 NOCRP

### PR DESCRIPTION
**Pull request by codyseibert**
Converting the notice of change to remote proceeding behind a websocket to allow for longer execution times.

In addition, we added the ability to mimic the asynchronous nature of a real API Gateway while running code locally.  This led us to discover places where our tests depended on data coming back from the async endpoints when, in reality, nothing would be coming back.

### CODE REVIEW

File: shared/src/business/useCases/trialSessions/updateTrialSessionInteractor.js [hunk 1]
```diff
@@ -188,31 +175,35 @@ exports.updateTrialSessionInteractor = async (
...
-      message: {
-        action: 'notice_generation_complete',
-        docketEntryId: serviceInfo.docketEntryId,
-        hasPaper: serviceInfo.hasPaper,
-        pdfUrl,
-      },
-      userId: user.userId,
-    });
+  if (trialSession.swingSession && trialSession.swingSessionId) {
+    newTrialSessionEntity.setAsSwingSession(trialSession.swingSessionId);
```
Comment (rachelschneiderman) [inline shared/src/business/useCases/trialSessions/updateTrialSessionInteractor.js:187]: Out of curiosity, why was this added?

File: web-api/src/lambdaWrapper.js [hunk 2]
```diff
@@ -26,21 +28,29 @@ exports.lambdaWrapper = lambda => {
       queryStringParameters: req.query,
     };
 
+    if (shouldMimicApiGateway) {
```
Comment (rachelschneiderman) [inline web-api/src/lambdaWrapper.js]: Nit: It seems like this condition is checked a few times for different reasons, is there a way they could be combined/streamlined?

File: web-api/src/lambdaWrapper.js [hunk 3]
```diff
@@ -26,21 +28,29 @@ exports.lambdaWrapper = lambda => {
...
+      res.status(204).send('');
```
Comment (rachelschneiderman) [inline web-api/src/lambdaWrapper.js:39]: Are there unit tests that can be written around these newly added lines that help document why they were needed/added?

File: web-client/integration-tests/utils/manuallyAddCaseToTrial.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
+// import { wait } from '../helpers';
```
Comment (rachelschneiderman) [inline web-client/integration-tests/utils/manuallyAddCaseToTrial.js]: Can these comments be removed?

File: web-client/integration-tests/utils/manuallyAddCaseToTrial.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+// import { withAppContextDecorator } from '../../src/withAppContext';
+
+export const manuallyAddCaseToTrial = cerebralTest => {
```
Comment (rachelschneiderman) [inline web-client/integration-tests/utils/manuallyAddCaseToTrial.js:1]: It looks like there are a couple other functions that also manually add a case to a trial session:

`docketClerkManuallyAddsCaseToTrialSessionWithoutNote`
`petitionsClerkManuallyAddsCaseToTrial`
`petitionsClerkManuallyAddsCaseToTrialWithoutJudge`

Could any of those been reused?

File: web-client/integration-tests/petitionsClerkSetsRemoteTrialSessionCalendar.test.js [hunk 6]
```diff
@@ -69,23 +79,12 @@ describe('petitions clerk sets a remote trial session calendar', () => {
       loginAs(cerebralTest, 'petitionsclerk@example.com');
       cerebralTest.casesReadyForTrial = [];
       petitionsClerkCreatesNewCase(cerebralTest, fakeFile, trialLocation);
-      petitionsClerkManuallyAddsCaseToTrial(cerebralTest);
+      manuallyAddCaseToTrial(cerebralTest);
     });
   });
 
-  describe('petitions clerk sets calendar for trial session', () => {
-    petitionsClerkViewsNewTrialSession(cerebralTest);
```
Comment (rachelschneiderman) [inline web-client/integration-tests/petitionsClerkSetsRemoteTrialSessionCalendar.test.js:77]: Why were these test steps removed?

File: web-client/src/presenter/actions/CaseDetail/setDefaultUnsealDocketEntryModalStateAction.test.js [hunk 7]
```diff
@@ -0,0 +1,30 @@
+import { applicationContextForClient as applicationContext } from '../../../../../shared/src/business/test/createTestApplicationContext';
+import { presenter } from '../../presenter-mock';
+import { runAction } from 'cerebral/test';
+import { setDefaultUnsealDocketEntryModalStateAction } from './setDefaultUnsealDocketEntryModalStateAction';
+
+describe('setDefaultUnsealDocketEntryModalStateAction', () => {
```
Comment (rosiehsmith) [inline web-client/src/presenter/actions/CaseDetail/setDefaultUnsealDocketEntryModalStateAction.test.js:6]: Thanks for adding test coverage here!

File: web-api/src/lambdaWrapper.js [hunk 8]
```diff
@@ -26,21 +28,29 @@ exports.lambdaWrapper = lambda => {
...
+        ...response.headers,
+        'X-Terminal-User': isTerminalUser,
+        ...exports.headerOverride,
+      });
+    }
 
-    if (
+    if (shouldMimicApiGateway) {
+      // Do nothing
+    } else if (
```
Comment (halprin) [inline web-api/src/lambdaWrapper.js]: Looks like the `else if` part of this `if` statement can be combined with the `if` statement above.

File: web-api/src/app.js [hunk 9]
```diff
@@ -1011,7 +1011,10 @@ app.get(
   );
   app.get('/trial-sessions', lambdaWrapper(getTrialSessionsLambda));
   app.post('/trial-sessions', lambdaWrapper(createTrialSessionLambda));
-  app.put('/trial-sessions', lambdaWrapper(updateTrialSessionLambda));
+  app.put(
+    '/async/trial-sessions',
+    lambdaWrapper(updateTrialSessionLambda, { isAsync: true }),
```
Comment (halprin) [inline web-api/src/app.js:1016]: Now that you added this `isAsync` option, shouldn't we update all of the other async endpoints in this file to have this option too?

File: web-client/src/presenter/actions/TrialSession/setPrintPaperDoneUrlAction.js [hunk 10]
```diff
@@ -0,0 +1,17 @@
+import { state } from 'cerebral';
+
+/**
+ * sets the printPaperDoneUrl onto the state used for knowing where to redirect when pressing done on the PrintPaperService component.
+ *
+ * @param {object} providers the providers object
+ * @param {object} providers.store the cerebral store used for setting the state.eligibleCases
```
Comment (rachelschneiderman) [inline web-client/src/presenter/actions/TrialSession/setPrintPaperDoneUrlAction.js]: Nit: docs need updating

File: web-api/src/lambdaWrapper.js [hunk 11]
```diff
@@ -10,8 +10,10 @@ exports.headerOverride = {
   'X-Content-Type-Options': 'nosniff',
 };
 
-exports.lambdaWrapper = lambda => {
+exports.lambdaWrapper = (lambda, options = {}) => {
   return async (req, res) => {
+    const shouldMimicApiGateway =
```
Comment (halprin) [inline web-api/src/lambdaWrapper.js]: `shouldMimicApiGateway` seems very generic.  In which way should it mimic the real AWS API Gateway?  In this case it wants to mimic the async-ness of it.  I would call it for what it is: something like `isAsyncNotInProduction`.

.... Oh wait.  this `NODE_ENV` environment variable.  That isn't the environment we deployed to, right?  I.e. not `exp3`, `test`, or `staging`.

**Review by halprin — COMMENTED** [review]
Not done with the review, but here are some comments thus far.

File: shared/src/business/useCases/trialSessions/updateTrialSessionInteractor.js [hunk 12]
```diff
@@ -188,31 +175,35 @@ exports.updateTrialSessionInteractor = async (
...
     applicationContext,
     trialSessionToUpdate: newTrialSessionEntity.validate().toRawObject(),
   });
 
+  await applicationContext.getNotificationGateway().sendNotificationToUser({
+    applicationContext,
+    message: {
+      action: 'update_trial_session_complete',
+      hasPaper: serviceInfo?.hasPaper,
+      pdfUrl,
```
Comment (rachelschneiderman) [inline shared/src/business/useCases/trialSessions/updateTrialSessionInteractor.js:200]: Seems like this logic was moved from `updateTrialSessionAction` to this interactor, if it's staying can test coverage be added for this new logic?

File: web-client/src/presenter/actions/navigateToMaintenanceAction.test.js [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+
+  beforeAll(() => {
+    routeStub = jest.fn();
+
+    presenter.providers.router = {
+      route: routeStub,
+    };
+  });
+
+  it('navigates to FilePetitionSuccess', async () => {
```
Comment (rachelschneiderman) [inline web-client/src/presenter/actions/navigateToMaintenanceAction.test.js]: Seems like this test name may have been a copy/pasta 😄

File: web-client/src/presenter/actions/setWaitingForResponseAction.js [hunk 14]
```diff
@@ -9,4 +9,5 @@ import { state } from 'cerebral';
 export const setWaitingForResponseAction = ({ store }) => {
   store.increment(state.progressIndicator.waitingForResponseRequests);
   store.set(state.progressIndicator.waitingForResponse, true);
+  store.unset(state.progressIndicator.waitText);
```
Comment (rachelschneiderman) [inline web-client/src/presenter/actions/setWaitingForResponseAction.js]: Is there test coverage for this?

File: web-client/src/presenter/actions/setWaitingTextAction.js [hunk 15]
```diff
@@ -0,0 +1,13 @@
+import { state } from 'cerebral';
+
+/**
+ * sets the wait text on the spinner
+ *
+ * @param {string} waitText the string to show on the spinner
+ * @returns {object} the set waiting for response action
+ */
+export const setWaitingTextAction =
```
Comment (rachelschneiderman) [inline web-client/src/presenter/actions/setWaitingTextAction.js:9]: Missing test coverage

File: web-client/src/presenter/actions/setWaitingForResponseAction.js (hunk 14 above)
Comment (rachelschneiderman) [inline web-client/src/presenter/actions/setWaitingForResponseAction.js]: I'm curious why this line was added? It seems like this action is called immediately before we set the wait text

File: shared/src/business/useCases/trialSessions/updateTrialSessionInteractor.js (hunk 1 above)
Comment (codyseibert) [reply shared/src/business/useCases/trialSessions/updateTrialSessionInteractor.js:187]: We used to have the UI invoke a separate endpoint to set the trial session as a swing state.  This logic seemed like it was more of a backend concern than a front end concern, so instead this single endpoint will check to set it as a swing state or not when updating the trial session.

File: web-api/src/lambdaWrapper.js (hunk 2 above)
Comment (codyseibert) [reply web-api/src/lambdaWrapper.js]: I'm not sure it will be easy to do that.  I'm updating the code to include a lot of comment to explain the purpose of this flag.

File: web-client/integration-tests/utils/manuallyAddCaseToTrial.js (hunk 5 above)
Comment (codyseibert) [reply web-client/integration-tests/utils/manuallyAddCaseToTrial.js:1]: We tried using some of those helper files but they have expect statements coded directly into the code, which makes them coded for a specific usecase which fails when we try to use them.  Instead of adding more arguments to the helper function to ignore the expect statements, we went with a new function that just adds the case to the trial and doesn't expect anything.  This allows us to keep the expect statement in the test file itself instead of helpers failing because we are using them for the wrong purpose.

File: web-client/src/presenter/actions/setWaitingForResponseAction.js (hunk 14 above)
Comment (codyseibert) [reply web-client/src/presenter/actions/setWaitingForResponseAction.js]: so the reason we added this here is because when you do a XHR request, the spinner will show up and display the waitText if it was set from the action previously in the sequence.  So if you navigate to a new page and do a different XHR request, the text will stick around on the next time we display a spinner.  We only wanted to add this spinner text to this specific user flow and not to every spinner.

File: web-api/src/app.js (hunk 9 above)
Comment (codyseibert) [reply web-api/src/app.js:1016]: yes, but it will probably break a LOT of tests since right now every test expects every endpoint to just wait until done processing.  I personally would rather make a devex item to do this work.

File: web-api/src/lambdaWrapper.js (hunk 11 above)
Comment (codyseibert) [reply web-api/src/lambdaWrapper.js]: I renamed it to `shouldMimicApiGatewayAyncEndpoint` which I think is more descriptive I hope?  yeah, NODE_ENV will only be 'production' when it's deployed to a real AWS environment

File: web-api/src/lambdaWrapper.js (hunk 3 above)
Comment (codyseibert) [reply web-api/src/lambdaWrapper.js:39]: I added a lot of comments to explain the purpose of this local only flag

File: web-client/integration-tests/petitionsClerkSetsRemoteTrialSessionCalendar.test.js (hunk 6 above)
Comment (codyseibert) [reply web-client/integration-tests/petitionsClerkSetsRemoteTrialSessionCalendar.test.js:77]: Because when creating a remote trial session, the status goes to `Open`.  So we can't use that helper method because it expects the trial session to be a status of `New`

File: web-api/src/lambdaWrapper.js [hunk 16]
```diff
@@ -10,8 +10,15 @@ exports.headerOverride = {
   'X-Content-Type-Options': 'nosniff',
 };
 
-exports.lambdaWrapper = lambda => {
+exports.lambdaWrapper = (lambda, options = {}) => {
   return async (req, res) => {
+    // This flag was added because when invoking async endpoints on API gateway, the api gateway will return
+    // a 204 response with no body immediately.  This was causing discrepancies between how these endpoints
+    // ran locally and how they ran when deployed to an AWS environment.  We now turn this flag on
+    // whenever we are not running in product to mimic how API gateway async endpoints work.
```
Comment (halprin) [inline web-api/src/lambdaWrapper.js]:
```suggestion
    // whenever we are not deployed in AWS to mimic how API gateway async endpoints work.
```

Comment (halprin) [inline web-api/src/lambdaWrapper.js]: Thanks for adding the comments.  They do a good job of explaining the why!

File: web-client/integration-tests/utils/manuallyAddCaseToTrial.js (hunk 5 above)
Comment (halprin) [reply web-client/integration-tests/utils/manuallyAddCaseToTrial.js:1]: Since this doesn't have any `expects`, can you call this function in the aforementioned functions so we don't repeat this logic?

Comment (rachelschneiderman) [reply web-client/integration-tests/utils/manuallyAddCaseToTrial.js:1]: Gotcha, makes sense. Thanks!

File: web-client/integration-tests/petitionsClerkSetsRemoteTrialSessionCalendar.test.js (hunk 6 above)
Comment (rachelschneiderman) [reply web-client/integration-tests/petitionsClerkSetsRemoteTrialSessionCalendar.test.js:77]: I am concerned we lost test coverage in removing these steps, or are the lines that were removed functionality that no longer exists in the system as of implementing this story?

File: web-client/src/presenter/actions/setWaitingForResponseAction.js (hunk 14 above)
Comment (rachelschneiderman) [reply web-client/src/presenter/actions/setWaitingForResponseAction.js]: So is that why it's not unset in the `unsetWaitingForResponseAction`?

File: web-client/integration-tests/petitionsClerkSetsRemoteTrialSessionCalendar.test.js (hunk 6 above)
Comment (codyseibert) [reply web-client/integration-tests/petitionsClerkSetsRemoteTrialSessionCalendar.test.js:77]: I'll make a helper called `petitionsClerkViewsOpenTrialSession` which should make this make more sense

**Review by halprin — COMMENTED** [review]
These are a little late.  Hopefully they still make sense.

File: web-client/src/presenter/actions/TrialSession/setPrintPaperDoneUrlAction.test.js [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+      },
+      props: {
+        trialSessionId: 'abc',
+      },
+      state: {
+        printPaperDoneUrl: null,
+      },
+    });
+
+    expect(result.state.printPaperDoneUrl).toEqual('/trial-session-detail/abc');
```
Comment (halprin) [inline web-client/src/presenter/actions/TrialSession/setPrintPaperDoneUrlAction.test.js]: Nit: extract `abc` into a variable that is used in running the test and in the expect.

File: web-client/src/presenter/actions/setWaitingForResponseAction.js (hunk 14 above)
Comment (halprin) [reply web-client/src/presenter/actions/setWaitingForResponseAction.js]: Would this be a good time to use some inline comments to explain _why_ we unset this?  I'll defer to others who understand the implications more than I.

File: web-client/src/presenter/actions/setWaitingTextAction.test.js [hunk 18]
```diff
@@ -0,0 +1,20 @@
+import { presenter } from '../presenter-mock';
+import { runAction } from 'cerebral/test';
+import { setWaitingTextAction } from './setWaitingTextAction';
+
+describe('setWaitingTextAction', () => {
+  it('should set the wait text when invoked with a string', async () => {
+    const EXPECTED_WAIT_TEXT = 'This is the wait text';
```
Comment (halprin) [inline web-client/src/presenter/actions/setWaitingTextAction.test.js:7]: Nice!

**Review by halprin — COMMENTED** [review]
Another batch.

File: web-client/src/presenter/actions/setWaitingForResponseAction.js (hunk 14 above)
Comment (codyseibert) [reply web-client/src/presenter/actions/setWaitingForResponseAction.js]: we refactored this to be placed in the unsetWaitingForResponseAction since we think it makes more sense there during cleanup.

Comment (rachelschneiderman) [reply web-client/src/presenter/actions/setWaitingForResponseAction.js]: I think I'd prefer a named test if possible

**Review by rachelschneiderman — APPROVED** [review]

**Comment by halprin** [comment]
@codyseibert, fixed the test.  Tomorrow I'd like to discuss the idea of removing `setPractitionerDetailAction` from the `success` path after calling `updatePractitionerUserAction` in `submitUpdatePractitionerUserSequence.js`.  Similarly, we should discuss removing passing the `barNumber` and `practitionerDetail` in the success path in `updatePractitionerUserAction.js` since it isn't used.  And really we might want to discuss removing the return statement at the bottom of `updatePractitionerUserInteractor.js` too.

File: shared/src/business/useCases/practitioners/updatePractitionerUserInteractor.js [hunk 19]
```diff
@@ -174,8 +174,4 @@ exports.updatePractitionerUserInteractor = async (
     },
     userId: requestUser.userId,
   });
-
-  return new Practitioner(updatedUser, { applicationContext })
```
Comment (rachelschneiderman) [inline shared/src/business/useCases/practitioners/updatePractitionerUserInteractor.js:178]: I believe I understand why this is being removed but I am concerned as this is changing currently accepted/deployed functionality. What is the reason for this change?

Comment (halprin) [reply shared/src/business/useCases/practitioners/updatePractitionerUserInteractor.js:178]: Talked this over in Zoom.

**Comment by halprin** [comment]
I believe we're done here.  The latest changes since about this time yesterday were all associated with adding the ability to mimic the asynchronous nature of the real API Gateway while running code locally.  This led us to discover places where our tests depended on data coming back from the async endpoints when, in reality, nothing would be coming back.  This also led us to add waits until after the progress indicator would disappear to some tests.

**Review by TomElliottFlexion — APPROVED** [review]

**Review by rosiehsmith — APPROVED** [review]
